### ISSUE TITLE
Use of sample shadow sets reported property automatically based on desired property

### ISSUE THREAD

**Issue by SidneyNiccolson** [awslabs/aws-iot-device-client#358]
**Describe the bug**

I'm relatively new to the Named Shadow Feature of iot-device-client. From my understanding of IoT shadows, a device typically only reports the "reported" state as stated here [ref](https://docs.aws.amazon.com/iot/latest/developerguide/iot-device-shadows.html)

-     Devices should write only to the reported property of the shadow state when communicating state data to the shadow.
-     Apps and other cloud services should write only to the desired property when communicating state change requests to the device through the shadow.

When using the Named Shadow feature I noticed that when I send a new "desired" state from the cloud, it also automatically updates the "reported" value to the "desired" value. This is something I did not expect as I assume that only "desired" is updated until the device has handled that request.

**To Reproduce**

1. My iot-device-client configuration snippet:
```

      "config-shadow":  {
        "enabled":      false
      },
      "sample-shadow": {
        "enabled": true,
        "shadow-name": "poc-shadow",
        "shadow-input-file": "/home/pi/input_shadow/poc-shadow-in.json",
        "shadow-output-file": "/home/pi/output_shadow/poc-shadow-out.json"
      }
    }
```
2. I placed the following sample content in my poc-shadow-in.json:

```
{
      "color": "red",
      "power": "on"
}
```

3. the poc-shadow-out.json is empty in this phase
4. I started the aws-iot-device-client which resulted in the creation of the AWS shadow at AWS IoT core:
```
{
  "state": {
    "desired": {
      "color": "red",
      "power": "on"
    },
    "reported": {
      "color": "red",
      "power": "on"
    }
  }
}
```
5. After I wanted to mimic a cloud MQTT shadow update request via MQTT test client:
Topic: $aws/things/mythingname/shadow/name/poc-shadow/update
payload:
```
{
  "state": {
    "desired": {
      "color": "green",
      "power": "on"
    }
  }
}
```
6. After executing this request, it sets the "reported" state to the value, which is something I did not expect. Looking deeper I noticed that it does 2 modifications while looking at the poc-shadow-out.json:

```
{
	"previous":	{
		"state":	{
			"desired":	{
				"color":	"green",
				"power":	"on"
			},
			"reported":	{
				"color":	"red",
				"power":	"on"
			}
		},
		"metadata":	{
			"desired":	{
				"color":	{
					"timestamp":	1670444780
				},
				"power":	{
					"timestamp":	1670444780
				}
			},
			"reported":	{
				"color":	{
					"timestamp":	1670444352
				},
				"power":	{
					"timestamp":	1670444352
				}
			}
		},
		"version":	34
	},
	"current":	{
		"state":	{
			"desired":	{
				"color":	"green",
				"power":	"on"
			},
			"reported":	{
				"color":	"green",
				"power":	"on"
			}
		},
		"metadata":	{
			"desired":	{
				"color":	{
					"timestamp":	1670444780
				},
				"power":	{
					"timestamp":	1670444780
				}
			},
			"reported":	{
				"color":	{
					"timestamp":	1670444780
				},
				"power":	{
					"timestamp":	1670444352
				}
			}
		},
		"version":	35
	},
	"timestamp":	1670444780
}
```

**Expected behavior**

The "reported" state does not get updated unless in my example poc-shadow-in.json is changed. 

**Actual behavior**

Both "reported" and "desired" gets updated in the shadow object without touching poc-shadow-in.json. poc-shadow-in.json still contains the initial value. Only an aws-iot-device-client restart causes it to get back to the original value, but also updates the "desired" value.

**Logs**

BEFORE REQUEST TO CHANGE DESIRED VALUE
2022-12-07T20:19:12.404Z [INFO]  {Main.cpp}: Sample shadow is enabled
2022-12-07T20:19:12.404Z [INFO]  {Main.cpp}: Pub Sub is disabled
2022-12-07T20:19:12.404Z [INFO]  {Main.cpp}: Sensor Publish is disabled
2022-12-07T20:19:12.404Z [INFO]  {SharedCrtResourceManager.cpp}: Starting Device Client features.
2022-12-07T20:19:12.404Z [INFO]  {Main.cpp}: Client base has been notified that Jobs has started
2022-12-07T20:19:12.404Z [INFO]  {SampleShadowFeature.cpp}: Starting SampleShadow
2022-12-07T20:19:12.404Z [INFO]  {JobsFeature.cpp}: Running Jobs!
2022-12-07T20:19:12.505Z [INFO]  {Main.cpp}: Client base has been notified that SampleShadow has started
2022-12-07T20:19:12.567Z [INFO]  {JobsFeature.cpp}: No pending jobs are scheduled, waiting for the next incoming job
2022-12-07T20:19:12.632Z [INFO]  {SampleShadowFeature.cpp}: Stored the latest poc-shadow shadow document to local successfully

AFTER REQUEST TO CHANGE "DESIRED" VALUE
....
2022-12-07T20:26:20.319Z [INFO]  {SampleShadowFeature.cpp}: Stored the latest poc-shadow shadow document to local successfully
2022-12-07T20:26:20.421Z [INFO]  {SampleShadowFeature.cpp}: Stored the latest poc-shadow shadow document to local successfully --> Here it changes the "reported" value automatically without any intermittent check on input-file.

**Comment by RogerZhongAWS** [comment]
Generally, it looks like the shadow feature is working as intended. Because the sample-shadow feature is intended to be more of a demo than a legitimate use case, we want to demonstrate the ability to receive a desired state from the cloud and instantly sync our reported state to match the desired state without waiting for some other action to be performed. 

However, I have observed that upon device client startup, the shadow feature reads the JSON from the input file sets both the desired and reported states to those values. I am not sure if this specific behavior is correct, as you mentioned here:

> Devices should write only to the reported property of the shadow state when communicating state data to the shadow.

Thus, the fact that we set both desired and reported states seems contradictory.
And then, because the shadow gets created such that the desired and reported states are the same, any attempt to change the reported state via a write to the input file would be immediately overwritten by the shadow sync handler which changes it back to the desired state.

In conclusion:
- Yes, automatically updating the reported value is intentional.
- No, we do not want device client to write to the input file in the event that it updates the reported value. The input file is more of a primitive 'UI' for users to manually update reported state.
- And yes, the fact that we create a desired property upon device client startup seems to not make a lot of sense. If a shadow does not have a desired property, device client probably should not create it. Any updates to the input file --> reported state should stick until the desired state gets populated from the cloud side. The change seems to be as simple as a one line deletion. Will double check to make sure.

**Comment by SidneyNiccolson** [comment]
Thanks @RogerZhongAWS for your quick reply. 

> Generally, it looks like the shadow feature is working as intended. Because the sample-shadow feature is intended to be more of a demo than a legitimate use case, we want to demonstrate the ability to receive a desired state from the cloud and instantly sync our reported state to match the desired state without waiting for some other action to be performed.

> No, we do not want device client to write to the input file in the event that it updates the reported value. The input file is more of a primitive 'UI' for users to manually update reported state.

Ok that is a bit unfortunate as the implementation of IoT jobs is a really nice useful feature and we hope to use Shadows in our semi-production setting as well with aws-iot-device-client.  In this case we simply cannot easily make use of this feature. We originally wished to use the input-file to update reported state without having to also restart aws-iot-device-client. In return use INotify to listen to changes from the cloud on the output-file without automatically matching desired to reported. I imagine that for this we simply need to built or own logic in our application directly and there are samples around from the AWS SDK.

> And yes, the fact that we create a desired property upon device client startup seems to not make a lot of sense. If a shadow does not have a desired property, device client probably should not create it. Any updates to the input file --> reported state should stick until the desired state gets populated from the cloud side. The change seems to be as simple as a one line deletion. Will double check to make sure.

Ok, thanks.
